File: src/synthorg/api/controllers/artifacts.py
"""Artifact controller -- endpoints for artifact management, storage, and retrieval."""

from typing import Annotated, Any

from litestar import Controller, Request, Response, delete, get, post, put
from litestar.datastructures import State  # noqa: TC002
from litestar.enums import RequestEncodingType
from litestar.params import Body, Parameter

from synthorg.api.channels import CHANNEL_ARTIFACTS, publish_ws_event
from synthorg.api.dto import ApiResponse, CreateArtifactRequest, PaginatedResponse
from synthorg.api.errors import (
    ArtifactStorageFullApiError,
    ArtifactTooLargeApiError,
    NotFoundError,
)
from synthorg.api.guards import require_read_access, require_write_access
from synthorg.api.pagination import CursorLimit, CursorParam, paginate_cursor
from synthorg.api.path_params import QUERY_MAX_LENGTH, PathId
from synthorg.api.rate_limits.guard import per_op_rate_limit
from synthorg.api.ws_models import WsEventType
from synthorg.core.artifact import Artifact
from synthorg.core.enums import ArtifactType
from synthorg.core.types import NotBlankStr
from synthorg.engine.artifacts.service import ArtifactService
from synthorg.observability import get_logger
from synthorg.observability.events.persistence import (
    PERSISTENCE_ARTIFACT_CONTENT_MISSING,
    PERSISTENCE_ARTIFACT_FETCH_FAILED,
    PERSISTENCE_ARTIFACT_SAVE_FAILED,
    PERSISTENCE_ARTIFACT_STORAGE_DELETE_FAILED,
    PERSISTENCE_ARTIFACT_STORAGE_ROLLBACK_FAILED,
    PERSISTENCE_ARTIFACT_STORE_FAILED,
    PERSISTENCE_ARTIFACT_STORED,
)
from synthorg.persistence.errors import (
    ArtifactStorageFullError,
    ArtifactTooLargeError,
    PersistenceError,
    RecordNotFoundError,
)

logger = get_logger(__name__)


def _service(state: State) -> ArtifactService:
    """Build the per-request :class:`ArtifactService` instance."""
    return ArtifactService(repo=state.app_state.persistence.artifacts)


_SAFE_CONTENT_TYPES = frozenset(
    {
        "application/octet-stream",
        "application/json",
        "application/pdf",
        "application/xml",
        "application/zip",
        "application/gzip",
        "application/x-tar",
        "image/png",
        "image/jpeg",
        "image/gif",
        # image/svg+xml intentionally excluded -- SVG is an XML document
        # with full JavaScript execution capability (XSS risk).
        "image/webp",
        "text/plain",
        "text/csv",
        "text/xml",
        "text/markdown",
    }
)

TaskIdFilter = Annotated[
    NotBlankStr | None,
    Parameter(
        required=False,
        max_length=QUERY_MAX_LENGTH,
        description="Filter by originating task ID",
    ),
]

CreatedByFilter = Annotated[
    NotBlankStr | None,
    Parameter(
        required=False,
        max_length=QUERY_MAX_LENGTH,
        description="Filter by creator agent ID",
    ),
]

TypeFilter = Annotated[
    NotBlankStr | None,
    Parameter(
        required=False,
        query="type",
        max_length=QUERY_MAX_LENGTH,
        description="Filter by artifact type",
    ),
]


async def _save_metadata_with_rollback(
    service: ArtifactService,
    storage: Any,
    artifact_id: str,
    updated: Artifact,
) -> None:
    """Save updated artifact metadata, rolling back storage on failure.

    Args:
        service: Artifact service wrapping the persistence repo.
        storage: Artifact content storage backend.
        artifact_id: Artifact identifier.
        updated: Updated artifact model.

    Raises:
        PersistenceError: If the metadata save fails (after rollback attempt).
    """
    try:
        await service.save(updated)
    except PersistenceError as exc:
        logger.warning(
            PERSISTENCE_ARTIFACT_SAVE_FAILED,
            artifact_id=artifact_id,
            error=str(exc),
            note="metadata save failed, rolling back content",
        )
        try:
            await storage.delete(artifact_id)
        except Exception as cleanup_exc:
            logger.warning(
                PERSISTENCE_ARTIFACT_STORAGE_ROLLBACK_FAILED,
                artifact_id=artifact_id,
                error=str(cleanup_exc),
            )
        raise


class ArtifactController(Controller):
    """Controller for artifact listing, creation, deletion, and content storage."""

    path = "/artifacts"
    tags = ("artifacts",)

    @get(guards=[require_read_access])
    async def list_artifacts(  # noqa: PLR0913
        self,
        state: State,
        cursor: CursorParam = None,
        limit: CursorLimit = 50,
        task_id: TaskIdFilter = None,
        created_by: CreatedByFilter = None,
        type: TypeFilter = None,  # noqa: A002
    ) -> PaginatedResponse[Artifact] | Response[ApiResponse[None]]:
        """List artifacts with optional filters.

        Args:
            state: Application state.
            cursor: Opaque pagination cursor from the previous page.
            limit: Page size.
            task_id: Filter by originating task ID.
            created_by: Filter by creator agent ID.
            type: Filter by artifact type.

        Returns:
            Paginated list of artifacts, or 400 for invalid filters.
        """
        parsed_type: ArtifactType | None = None
        if type is not None:
            try:
                parsed_type = ArtifactType(type)
            except ValueError:
                valid = ", ".join(e.value for e in ArtifactType)
                return Response(
                    content=ApiResponse[None](
                        error=(
                            f"Invalid artifact type: {type!r}. Valid values: {valid}"
                        ),
                    ),
                    status_code=400,
                )

        artifacts = await _service(state).list_artifacts(
            task_id=task_id,
            created_by=created_by,
            artifact_type=parsed_type,
        )
        page, meta = paginate_cursor(
            artifacts,
            limit=limit,
            cursor=cursor,
            secret=state.app_state.cursor_secret,
        )
        return PaginatedResponse[Artifact](data=page, pagination=meta)

    @get("/{artifact_id:str}", guards=[require_read_access])
    async def get_artifact(
        self,
        state: State,
        artifact_id: PathId,
    ) -> Response[ApiResponse[Artifact]]:
        """Get an artifact by ID.

        Args:
            state: Application state.
            artifact_id: Artifact identifier.

        Returns:
            The artifact metadata, or 404 if not found.
        """
        artifact = await _service(state).get(artifact_id)
        if artifact is None:
            return Response(
                content=ApiResponse[Artifact](
                    error=f"Artifact {artifact_id!r} not found",
                ),
                status_code=404,
            )
        return Response(
            content=ApiResponse[Artifact](data=artifact),
            status_code=200,
        )

    @post(guards=[require_write_access])
    async def create_artifact(
        self,
        request: Request[Any, Any, Any],
        state: State,
        data: CreateArtifactRequest,
    ) -> Response[ApiResponse[Artifact]]:
        """Create a new artifact.

        Args:
            request: The incoming request.
            state: Application state.
            data: Artifact creation payload.

        Returns:
            The created artifact with generated ID.
        """
        artifact = await _service(state).create(
            artifact_type=data.type,
            path=data.path,
            task_id=data.task_id,
            created_by=data.created_by,
            description=data.description,
            content_type=data.content_type,
            project_id=data.project_id,
        )
        publish_ws_event(
            request,
            WsEventType.ARTIFACT_CREATED,
            CHANNEL_ARTIFACTS,
            {
                "artifact_id": artifact.id,
                "task_id": artifact.task_id,
                "created_by": artifact.created_by,
                "type": artifact.type.value,
            },
        )
        return Response(
            content=ApiResponse[Artifact](data=artifact),
            status_code=201,
        )

    @delete(
        "/{artifact_id:str}",
        guards=[require_write_access],
        status_code=200,
    )
    async def delete_artifact(
        self,
        request: Request[Any, Any, Any],
        state: State,
        artifact_id: PathId,
    ) -> Response[ApiResponse[None]]:
        """Delete an artifact and its stored content.

        Args:
            request: The incoming request.
            state: Application state.
            artifact_id: Artifact identifier.

        Returns:
            200 on success, 404 if not found.
        """
        service = _service(state)
        artifact = await service.get(artifact_id)
        if artifact is None:
            return Response(
                content=ApiResponse[None](
                    error=f"Artifact {artifact_id!r} not found",
                ),
                status_code=404,
            )
        # Delete storage content first -- if this fails, metadata still
        # exists so the inconsistency is detectable (vs. the reverse
        # order where metadata is gone but orphaned blob is invisible).
        storage = state.app_state.artifact_storage
        try:
            await storage.delete(artifact_id)
        except Exception as exc:
            logger.warning(
                PERSISTENCE_ARTIFACT_STORAGE_DELETE_FAILED,
                artifact_id=artifact_id,
                error=str(exc),
            )
        await service.delete(artifact_id)
        publish_ws_event(
            request,
            WsEventType.ARTIFACT_DELETED,
            CHANNEL_ARTIFACTS,
            {"artifact_id": artifact_id, "task_id": artifact.task_id},
        )
        return Response(
            content=ApiResponse[None](data=None),
            status_code=200,
        )

    @put(
        "/{artifact_id:str}/content",
        guards=[
            require_write_access,
            per_op_rate_limit(
                "artifacts.upload",
                max_requests=10,
                window_seconds=60,
                key="user",
            ),
        ],
        media_type="application/json",
    )
    async def upload_content(
        self,
        request: Request[Any, Any, Any],
        state: State,
        artifact_id: PathId,
        data: Annotated[
            bytes,
            Body(media_type=RequestEncodingType.MULTI_PART),
        ],
    ) -> Response[ApiResponse[Artifact]]:
        """Upload binary content for an artifact.

        Validates size limits before storing.

        Args:
            request: The incoming request.
            state: Application state.
            artifact_id: Artifact identifier.
            data: Binary content.

        Returns:
            Updated artifact metadata with size_bytes set.
        """
        service = _service(state)
        artifact = await service.get(artifact_id)
        if artifact is None:
            msg = f"Artifact {artifact_id!r} not found"
            logger.warning(
                PERSISTENCE_ARTIFACT_FETCH_FAILED,
                artifact_id=artifact_id,
                error_type="artifact_not_found",
                note="upload_content_target_missing",
            )
            raise NotFoundError(msg)

        storage = state.app_state.artifact_storage
        try:
            size = await storage.store(artifact_id, data)
        except ArtifactTooLargeError as exc:
            logger.warning(
                PERSISTENCE_ARTIFACT_STORE_FAILED,
                artifact_id=artifact_id,
                error_type=type(exc).__name__,
                error=str(exc),
                note="artifact_too_large",
            )
            raise ArtifactTooLargeApiError from exc
        except ArtifactStorageFullError as exc:
            logger.warning(
                PERSISTENCE_ARTIFACT_STORE_FAILED,
                artifact_id=artifact_id,
                error_type=type(exc).__name__,
                error=str(exc),
                note="artifact_storage_full",
            )
            raise ArtifactStorageFullApiError from exc

        updated = artifact.model_copy(
            update={
                "size_bytes": size,
                "content_type": (artifact.content_type or "application/octet-stream"),
            },
        )
        await _save_metadata_with_rollback(service, storage, artifact_id, updated)
        logger.info(
            PERSISTENCE_ARTIFACT_STORED,
            artifact_id=artifact_id,
            size_bytes=size,
        )
        publish_ws_event(
            request,
            WsEventType.ARTIFACT_CONTENT_UPLOADED,
            CHANNEL_ARTIFACTS,
            {
                "artifact_id": artifact_id,
                "size_bytes": size,
                "content_type": updated.content_type,
            },
        )
        return Response(
            content=ApiResponse[Artifact](data=updated),
            status_code=200,
        )

    @get(
        "/{artifact_id:str}/content",
        guards=[require_read_access],
        media_type="application/octet-stream",
    )
    async def download_content(
        self,
        state: State,
        artifact_id: PathId,
    ) -> Response:  # type: ignore[type-arg]
        """Download binary content for an artifact.

        Args:
            state: Application state.
            artifact_id: Artifact identifier.

        Returns:
            Binary content with appropriate content type, or JSON
            error on 404.
        """
        artifact = await _service(state).get(artifact_id)
        if artifact is None:
            return Response(
                content=ApiResponse[None](error="Artifact not found"),
                status_code=404,
                media_type="application/json",
            )

        storage = state.app_state.artifact_storage
        try:
            content = await storage.retrieve(artifact_id)
        except RecordNotFoundError:
            logger.warning(
                PERSISTENCE_ARTIFACT_CONTENT_MISSING,
                artifact_id=artifact_id,
            )
            return Response(
                content=ApiResponse[None](error="Artifact content not found"),
                status_code=404,
                media_type="application/json",
            )

        raw_ct = artifact.content_type or "application/octet-stream"
        fallback = "application/octet-stream"
        safe_ct = raw_ct if raw_ct in _SAFE_CONTENT_TYPES else fallback
        return Response(
            content=content,
            status_code=200,
            media_type=safe_ct,
            headers={"Content-Disposition": "attachment"},
        )

File: web/src/mocks/handlers/artifacts.ts
import { http, HttpResponse } from 'msw'
import type {
  createArtifact,
  getArtifact,
  listArtifacts,
} from '@/api/endpoints/artifacts'
import type { Artifact } from '@/api/types/artifacts'
import type { PaginatedResponse, PaginationMeta } from '@/api/types/http'
import {
  apiError,
  apiSuccess,
  emptyPage,
  paginatedFor,
  successFor,
  voidSuccess,
} from './helpers'

export function buildArtifact(overrides: Partial<Artifact> = {}): Artifact {
  return {
    id: 'artifact-default',
    type: 'code',
    path: 'src/default.ts',
    task_id: 'task-default',
    created_by: 'agent-default',
    description: 'Default artifact stub',
    project_id: null,
    content_type: 'text/plain',
    size_bytes: 0,
    created_at: '2026-04-19T00:00:00Z',
    ...overrides,
  }
}

// ── Storybook-facing named export: populated paginated list. ──
const mockArtifacts: Artifact[] = [
  {
    id: 'artifact-abc123',
    type: 'code',
    path: 'src/engine/coordinator.py',
    task_id: 'task-001',
    created_by: 'agent-eng-001',
    description: 'Coordinator implementation',
    project_id: 'proj-001',
    content_type: 'text/plain',
    size_bytes: 4096,
    created_at: '2026-03-30T10:00:00Z',
  },
  {
    id: 'artifact-def456',
    type: 'tests',
    path: 'tests/test_coordinator.py',
    task_id: 'task-001',
    created_by: 'agent-qa-001',
    description: 'Coordinator tests',
    project_id: 'proj-001',
    content_type: 'text/plain',
    size_bytes: 2048,
    created_at: '2026-03-30T11:00:00Z',
  },
  {
    id: 'artifact-ghi789',
    type: 'documentation',
    path: 'docs/coordinator.md',
    task_id: 'task-002',
    created_by: 'agent-eng-001',
    description: 'Coordinator documentation',
    project_id: null,
    content_type: 'text/markdown',
    size_bytes: 1024,
    created_at: '2026-03-30T12:00:00Z',
  },
]

const pagination: PaginationMeta = {
  total: mockArtifacts.length,
  offset: 0,
  limit: 50,
}

export const artifactsList = [
  http.get('/api/v1/artifacts', () => {
    const body: PaginatedResponse<Artifact> = {
      data: mockArtifacts,
      error: null,
      error_detail: null,
      success: true,
      pagination,
    }
    return HttpResponse.json(body)
  }),
  http.get('/api/v1/artifacts/:id', ({ params }) => {
    const artifact = mockArtifacts.find((a) => a.id === params.id)
    if (!artifact) {
      return HttpResponse.json(apiError('Artifact not found'), { status: 404 })
    }
    return HttpResponse.json(apiSuccess(artifact))
  }),
]

// ── Default test handlers: empty list and generic single-item lookups. ──
export const artifactsHandlers = [
  http.get('/api/v1/artifacts', () =>
    HttpResponse.json(paginatedFor<typeof listArtifacts>(emptyPage<Artifact>())),
  ),
  http.get('/api/v1/artifacts/:id', ({ params }) =>
    HttpResponse.json(
      successFor<typeof getArtifact>(buildArtifact({ id: String(params.id) })),
    ),
  ),
  http.post('/api/v1/artifacts', async ({ request }) => {
    const body = (await request.json()) as {
      type?: Artifact['type']
      path?: string
      task_id?: string
      created_by?: string
      description?: string
      content_type?: string
      project_id?: string | null
    }
    if (!body.type || !body.path || !body.task_id || !body.created_by) {
      return HttpResponse.json(apiError('Missing required fields'), {
        status: 400,
      })
    }
    return HttpResponse.json(
      successFor<typeof createArtifact>(
        buildArtifact({
          id: `artifact-${body.task_id}`,
          type: body.type,
          path: body.path,
          task_id: body.task_id,
          created_by: body.created_by,
          description: body.description ?? '',
          content_type: body.content_type ?? 'text/plain',
          project_id: body.project_id ?? null,
        }),
      ),
      { status: 201 },
    )
  }),
  http.delete('/api/v1/artifacts/:id', () => HttpResponse.json(voidSuccess())),
  http.get('/api/v1/artifacts/:id/content', ({ request }) => {
    const accept = request.headers.get('accept') ?? ''
    if (accept.includes('text')) {
      return new HttpResponse('default artifact content', {
        headers: { 'Content-Type': 'text/plain' },
      })
    }
    return new HttpResponse(new Blob(['default artifact content']), {
      headers: { 'Content-Type': 'application/octet-stream' },
    })
  }),
]
